notifications_utils/formatters.py

```diff
@@ -317,6 +317,42 @@ def get_action_link_image_url() -> str:
     return f'https://{img_env}-va-gov-assets.s3-us-gov-west-1.amazonaws.com/img/vanotify-action-link.png'
 
 
+def escape_whitespace_in_markdown_link(markdown: str) -> str:
+    """
+    Escapes all whitespace characters in the URL portion of Markdown links to prevent Markdown parsing issues.
+
+    This function ensures links like:
+        [Label](https://example.com/with spaces\tand\nnewlines)
+    become:
+        [Label](https://example.com/with%20spaces%09and%0Anewlines)
+
+    Notes:
+    - Only the URL portion is modified.
+    - Escapes: space, tab, newline, carriage return, vertical tab, form feed.
+    - Does not sanitize URLs for safety - only escapes whitespace to ensure Markdown compatibility.
+    - Avoids use of urllib to minimize over-escaping and maximize performance.
+
+    Args:
+        markdown (str): Raw markdown text containing links.
+
+    Returns:
+        str: Markdown with whitespace in link URLs percent-encoded.
+    """
+    def escape_whitespace(match: re.Match[str]) -> str:
+        label = match.group(1)
+        raw_url = match.group(2)
+
+        # Encode whitespace characters as a percent-encoded uppercase hex values
+        # (e.g., ' ' -> '%20')
+        escaped_url = re.sub(r'\s', lambda m: f'%{ord(m.group(0)):02X}', raw_url)
+
+        return f'[{label}]({escaped_url})'
+
+    # Match Markdown-style links: [label](url)
+    # URL match is non-greedy to support URLs with parentheses or placeholders like ((tag))
+    return re.sub(r'\[([^\]]+)\]\((.+?)\)', escape_whitespace, markdown)
```

Comment:
This also matches markdown for an image, which differs only in that it has a leading exclamation.  We don't need to substitute for images because they get deleted by custom mistune, so that would be a waste of computation.  I suggest prepending a negative look-behind to your regex so we don't match images.

You also don't need to alter the link text.  Use a positive lookbehind for that.
```suggestion
    return re.sub(r'(?<!!)\[([^\]]+)\]\((.+?)\)', escape_whitespace, markdown)
```

Reply:
Image markup needs to be escaped as well or mistune won't be able to parse and strip the images later on.